numpy/_core/src/multiarray/arraytypes.c.src

```diff
@@ -4280,7 +4281,7 @@ PyArray_DescrFromType(int type)
         PyErr_SetString(PyExc_ValueError,
                 "Invalid data-type for array");
     }
-    else {
+    else if (!is_stringdtype) {
```

Comment:
Can't say I like this control flow, but also don't want to iterate on this PR.

So thanks, let's just put it in.

@ngoldbaum I might rerun valgrind, but let's see (as you know running it this way, the full test suite takes 2 days rather than 2 minutes :)).